src/backend/distributed/metadata/node_metadata.c

```diff
@@ -1131,16 +839,73 @@ PrimaryNodeForGroup(int32 groupId, bool *groupContainsNodes)
 
 
 /*
- * ActivateNodeList iterates over the nodeList and activates the nodes.
- * Some part of the node activation is done parallel across the nodes,
- * such as syncing the metadata. However, reference table replication is
- * done one by one across nodes.
+ * NontransactionalMetadataSyncFailureHandler handles failure during nontransactional
+ * metadata sync.
+ */
+static void
+NontransactionalMetadataSyncFailureHandler(MetadataSyncContext *context)
+{
+	Assert(context->transactionMode == METADATA_SYNC_NON_TRANSACTIONAL);
+	Assert(!InMetadataSyncTestMode(context));
+
+	/*
+	 * set flags to false for the metadata nodes in given metadataSyncContext to
+	 * prevent an inconsistent intermediate state in case nontransactional
+	 * metadata sync fails before we activate those nodes.
+	 * 1) sets isactive to false, which means no query is allowed from any node
+	 * 2) sets metadatasynced to false, which means no DDL
+	 *   will be synced to the node and no query can be executed from that node to
+	 *   other nodes.
+	 */
+	bool activeAndSynced = false;
+	WorkerNode *node = NULL;
+	foreach_ptr(node, context->activatedWorkerNodeList)
+	{
+		SetWorkerColumnViaMetadataContext(context, node, Anum_pg_dist_node_isactive,
```

Comment:
I'm really not sure `SetWorkerColumnViaMetadataContext` is the right approach here. Any of the nodes could be down, and we could be here because of that.

I think we should only try setting these locally like:
```C
			node = SetWorkerColumnLocalOnly(node, Anum_pg_dist_node_isactive,
											BoolGetDatum(true));
			node = SetWorkerColumnLocalOnly(node, Anum_pg_dist_node_metadatasynced,
											BoolGetDatum(true));
			node = SetWorkerColumnLocalOnly(node, Anum_pg_dist_node_hasmetadata,
											BoolGetDatum(true));
```

We should aim for coordinator to have the right metadata. The operator should fix the workers later.

What do you think?

Reply:
One another option could be to follow what citus_update_node does:
```C
	if (UnsetMetadataSyncedForAllWorkers())
	{
		TriggerNodeMetadataSyncOnCommit();
	}
```

Where we set metadata locally, and let bg daemon try fixing it. 

Not sure yet which one is better. Any thoughts?

Reply:
We deactivate nodes in a separate connection to localhost before locally taking Exclusive lock on pg_dist_node. If local transaction fails, we are left with isactive and metadatasynced set to false for those nodes.

Reply:
@aykut-bozkurt can you pleease summarize our in-person meeting with Marco on this topic here for future reference

Reply:
We need to make sure that we set `isactive` and `metadatasynced` flags to false at the start of the `nontransactional metadata sync` because we donot want to send queries and ddls to those nodes until they are synced properly in case of a failure in the middle of the operations.

We could not set those flags in local transaction as any change would rollback if the local transaction fails. Thus, we set those flags to false in separate connection to `localhost` at the start of the `nontransactional` mode.